# Содержание

1. [Описание проекта](#start)
 * [Описание данных](#data)
 * [Задачи](#tasks)
 * [Цель исследования](#goal)
2. [Исследование таблиц](#table_research)
 * [Таблица books](#books)
 * [Таблица authors](#authors)
 * [Таблица publishers](#publishers)
 * [Таблица reviews](#reviews)
3. [Расчет количества книг, выпущенных после 1 января 2000 года](#number_of_books)
4. [Расчет количества обзоров и средней оценки для каждой книги](#overview_and_average_rating)
5. [Определение издательства, которое выпустило наибольшее число книг толще 50 страниц](#max_books)
6. [Определение автора с самой высокой средней оценкой книг (с учетом только книг с 50 и более оценками)](#average_book_rating)  
7. [Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок](#average_review)  
8. [Общий вывод](#conclusion)
9. [Рекомендации](#recommendations)

# Проект по SQL

<a id="start"></a> 
##  Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<a id="data"></a>
### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

<a id="tasks"></a>
### Задачи

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

<a id="goal"></a>
### Цель исследования

Цель исследования - проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

<a id="table_research"></a>
## Исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
# создаем функцию для чтения таблиц и присвоения им имен
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [4]:
query = '''
SELECT * FROM books
'''
books = pd.io.sql.read_sql(query, con = engine)
query_1 = '''
SELECT * FROM authors
'''

authors = pd.io.sql.read_sql(query_1, con = engine)
query_2 = '''
SELECT * FROM publishers
'''

publishers = pd.io.sql.read_sql(query_2, con = engine)
query_3 = '''
SELECT * FROM ratings
'''

ratings = pd.io.sql.read_sql(query_3, con = engine)
query_4 = '''
SELECT * FROM reviews
'''

reviews = pd.io.sql.read_sql(query_4, con = engine)

<a id="books"></a>
### Таблица books

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

In [5]:
pd.io.sql.read_sql('''
SELECT 
    *
FROM books
LIMIT 5
''', con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


Таблица содержит информацию о 1000 книг.

<a id="authors"></a>
### Таблица authors

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

In [7]:
pd.io.sql.read_sql('''
SELECT 
    *
FROM authors
LIMIT 5
''', con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


Таблица содержит информацию о 636 авторах.

<a id="publishers"></a>
### Таблица publishers

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

In [9]:
pd.io.sql.read_sql('''
SELECT 
    *
FROM publishers
LIMIT 5
''', con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [10]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


Таблица содержит информацию о 340 издательствах.

<a id="ratings"></a>
### Таблица ratings

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

In [11]:
pd.io.sql.read_sql('''
SELECT 
    *
FROM ratings
LIMIT 5
''', con = engine)


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


Таблица содержит информацию о 6456 оценках книг.

<a id="reviews"></a>
### Таблица reviews

Содержит данные о пользовательских обзорах на книги:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя пользователя, написавшего обзор;
- text — текст обзора.

In [13]:
pd.io.sql.read_sql('''
SELECT 
    *
FROM reviews
LIMIT 5
''', con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [14]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Таблица содержит информацию  о 2793 обзорах на книги.

<a id="number_of_books"></a>
## Расчет количества книг, выпущенных после 1 января 2000 года

In [15]:
pd.io.sql.read_sql('''
SELECT 
    COUNT(DISTINCT book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
''', con = engine)

cnt
0  819

**Вывод**

В базе данных содержится информация о 819 книгах, дата публикации которых позже 1 января 2000 года.

<a id="overview_and_average_rating"></a>
## Расчет количества обзоров и средней оценки для каждой книги

In [16]:
pd.io.sql.read_sql('''
SELECT 
    books.title AS title,
    books.book_id AS id,
    AVG(ratings.rating) AS avg_rating,
    COUNT(DISTINCT reviews.text) AS cnt_reviews
FROM
    books
INNER JOIN ratings ON ratings.book_id = books.book_id
INNER JOIN reviews ON reviews.book_id = books.book_id
GROUP BY
     books.title,
     books.book_id
order by
     cnt_reviews DESC
limit 20
''', con = engine)

title   id  avg_rating  \
0                             Twilight (Twilight  #1)  948    3.662500   
1                                       The Alchemist  627    3.789474   
2                           Outlander (Outlander  #1)  497    4.125000   
3                 The Hobbit  or There and Back Again  750    4.125000   
4                                     Eat  Pray  Love  207    3.395833   
5                                      The Book Thief  656    4.264151   
6   The Curious Incident of the Dog in the Night-Time  695    4.081081   
7                              The Catcher in the Rye  673    3.825581   
8                           The Giver (The Giver  #1)  733    3.750000   
9                                    The Glass Castle  734    4.206897   
10  The Lightning Thief (Percy Jackson and the Oly...  779    4.080645   
11                                           The Road  854    3.772727   
12                                Water for Elephants  963    3.977273   
13             The Da Vinci Code (Robert Langdon  #2)  696    3.830508   
14  Harry Potter and the Chamber of Secrets (Harry...  299    4.287500   
15  Harry Potter and the Prisoner of Azkaban (Harr...  302    4.414634   
16     A Clash of Kings  (A Song of Ice and Fire  #2)   15    4.391304   
17                Extremely Loud and Incredibly Close  232    3.416667   
18  Freakonomics: A Rogue Economist Explores the H...  267    3.909091   
19                                            Dracula  194    3.296296   

    cnt_reviews  
0             7  
1             6  
2             6  
3             6  
4             6  
5             6  
6             6  
7             6  
8             6  
9             6  
10            6  
11            6  
12            6  
13            6  
14            6  
15            6  
16            5  
17            5  
18            5  
19            5

**Вывод**

Для 7 книг отсуствуют рейтинг и обзор, максимальное кол-во обзоров у книги Twilight (Twilight #1). Достаточно много книг с оценкой 5, но у них всего 2 обзора.

<a id="max_books"></a>
## Определение издательства, которое выпустило наибольшее число книг толще 50 страниц

In [17]:
pd.io.sql.read_sql('''
SELECT
    publishers.publisher,
    COUNT(book_id) AS books_cnt
FROM
    books
    LEFT JOIN publishers ON publishers.publisher_id = books.publisher_id
WHERE 
    num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_cnt DESC
LIMIT 1
''', con = engine)

publisher  books_cnt
0  Penguin Books         42

**Вывод**

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги. Это достаточно старое и серьезное издательство, поэтому неудивительно, что оно на первом месте.

<a id="average_book_rating"></a>
## Определение автора с самой высокой средней оценкой книг (с учетом только книг с 50 и более оценками)

In [18]:
pd.io.sql.read_sql('''
SELECT
    authors.author,
    AVG(subquery_ratings.avg_rating) AS avg_rating
FROM
    books
    LEFT JOIN (
        SELECT 
            book_id,
            AVG(rating) AS avg_rating,
            COUNT(rating_id) AS rating_cnt
        FROM
            ratings
        GROUP BY
            book_id
    ) AS subquery_ratings ON subquery_ratings.book_id = books.book_id
    LEFT JOIN authors ON authors.author_id = books.author_id
WHERE
    rating_cnt >= 50
GROUP BY
    author
ORDER BY
    avg_rating DESC
LIMIT 1
''', con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.283844

**Вывод**

Самая высокая средняя оценка книг у автора Д. Роулинг. Это довольно ожидаемый результат - серия книг о Гарри Поттере действительно пользуется популярностью по всему миру. Средняя оценка ее книг составляет 4.28 балла.

<a id="average_review"></a>
## Расчет среднего количества обзоров от пользователей, которые поставили больше 50 оценок

In [19]:
pd.io.sql.read_sql('''
WITH
    ratings_subquery AS (
        SELECT
            ratings.username,
            COUNT(rating_id) AS ratings_cnt
        FROM
            ratings
        GROUP BY
            ratings.username
        HAVING
            COUNT(rating_id) > 50
        ),
    reviews_subquery AS (
        SELECT
            reviews.username,
            COUNT(review_id) AS reviews_cnt
        FROM
            reviews
        GROUP BY
            reviews.username
    )
SELECT
    AVG(reviews_cnt) as avg_reviews_cnt
FROM
    ratings_subquery
    LEFT JOIN reviews_subquery ON reviews_subquery.username = ratings_subquery.username
''', con = engine)

avg_reviews_cnt
0        24.333333

**Вывод**

Найдена информация об активных пользователях, которые поставили более 50 оценок. В среднем такие пользователи оставляют 24 обзора.

В ходе данного проекта была проанализирована база данных крупного сервиса для чтения книг по подписке. В ней содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. В резулате анализа получены следующие результаты:

1. В БД нового сервиса содержится информация о 1000 книг. Для крупного сервиса это не так много.
2. В БД содержится информация о 636 авторах, 340 издательствах, 6456 оценках книг и о 2793 обзорах на книги.
3. Порядка 82% книг выпущены за последние 20 лет (819).
4. Для большинства книг в БД есть оценки и обзоры (99%).
5. Максимальное количество обзоров не слишком велико - 7, оно пришлось на книгу Сумерки.
6. Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books. Оно выпустило 42 книги, что составляет 4% от выборки.
7. Самые высокие оценки у автора Д. Роулинг. Средняя оценка ее книг составляет 4.28 балла.


<a id="recommendations"></a>
## Рекомендации

1. Предлагать подписчикам свежие книги;
2. Предлагать подписчикам популярные книги, но интересные пользователю;
3. Предлагать книги издательств Penguin Books;
4. Акцентировать внимание пользователей на книгах авторов J.K. Rowling и Mary GrandPré;